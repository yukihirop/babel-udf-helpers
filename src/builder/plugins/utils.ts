import { NodePath } from '@babel/traverse';

// MEMO:
// generate <path> used by path.get(<path>)
export function makePath(path: NodePath): string {
  const parts = [];
  // If the parent exists, set it to path and execute it.Execution ends when there are no parents
  for (; path.parentPath; path = path.parentPath) {
    parts.push(path.key);
    if (path.inList) parts.push(path.listKey);
  }

  return parts.reverse().join('.');
}
