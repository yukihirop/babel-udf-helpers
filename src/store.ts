import { UDFHelpers } from './types';

export let helpers = Object.create(null) as UDFHelpers;

export function clearAll() {
  clearHelpers();
}

export function clearHelpers() {
  helpers = Object.create(null) as UDFHelpers;
}
