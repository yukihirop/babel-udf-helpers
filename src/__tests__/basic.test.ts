import { clearAll as clearAllHelpers } from '../store';
import { writeFile } from 'fs';
import printer from './printer';
import { inputFixturePath, outputFixturePath } from './utils';

beforeEach(() => {
  jest.spyOn(console, 'log').mockImplementation();
  jest.spyOn(console, 'warn').mockImplementation();
});

afterEach(() => {
  jest.clearAllMocks();
  clearAllHelpers();
});

describe('addUDFHelper', () => {
  describe('basic', () => {
    const type = 'basic';

    const dirs = ['declaration', 'dependencies'];

    for (const dir of dirs) {
      test(dir, () => {
        const helpers = require(inputFixturePath([type, dir]));
        // @ts-ignore
        const programFunc = (pass) => pass.addUDFHelper(dir);
        const code = printer({ helpers, programFunc });

        expect(code).toMatchSnapshot();
        writeFile(outputFixturePath([type, dir]), code || '', () => {});
      });
    }

    test('rename', () => {
      const dir = 'rename';
      const helpers = require(inputFixturePath([type, dir]));
      const content = `function _rename() { return "path.replaceWith" }`;
      const programFunc = (pass) => pass.addUDFHelper(dir);
      // @ts-ignore
      const code = printer({ content, helpers, programFunc });

      expect(code).toMatchSnapshot();
      writeFile(outputFixturePath([type, dir]), code || '', () => {});
    });
  });
});

describe('listUDFHelper', () => {
  describe('basic', () => {
    const type = 'basic';

    test('officialMix', () => {
      const dir = 'officialMix';
      const helpers = require(inputFixturePath([type, dir]));
      const programFunc = (pass) => {
        console.log(pass.listUDFHelper());
      };

      // @ts-ignore
      printer({ helpers, programFunc });
      expect((console.log as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
    });
  });
});
