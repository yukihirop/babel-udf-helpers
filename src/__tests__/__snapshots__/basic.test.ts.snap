// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addUDFHelper basic declaration 1`] = `"function _declaration() { console.log(\\"declaration\\"); }"`;

exports[`addUDFHelper basic dependencies 1`] = `
"function _dependencies() { cild1() || _child2(); }

;

function _child2() { _grandchild(); }

;

function _grandchild() { return \\"grandchild\\"; }

;

function _child() { return \\"child1\\"; }

;"
`;

exports[`addUDFHelper basic rename 1`] = `
"function _rename2() { console.log(\\"rename\\"); }

function _rename() {
  return \\"path.replaceWith\\";
}"
`;

exports[`listUDFHelper basic officialMix 1`] = `
Object {
  "available": Array [
    "officialMix",
    "child1",
    "child2",
    "grandchild",
  ],
  "unavailable": Array [
    "objectWithoutPropertiesLoose",
    "objectWithoutProperties",
  ],
}
`;
